clusterloader2/pkg/measurement/common/chaos_monkey_measurement.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/api"
+	"k8s.io/perf-tests/clusterloader2/pkg/chaos"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+)
+
+const (
+	chaosMonkeyMeasurementName = "ChaosMonkey"
+)
+
+func init() {
+	create := func() measurement.Measurement {
+		return &chaosMonkeyMeasurement{}
+	}
+	if err := measurement.Register(chaosMonkeyMeasurementName, create); err != nil {
+		klog.Fatalf("Cannot register %s: %v", chaosMonkeyMeasurementName, err)
+	}
+}
+
+type chaosMonkeyMeasurement struct {
+	api.NodeFailureConfig
+	stopChannel          chan struct{}
+	chaosMonkey          *chaos.Monkey
+	chaosMonkeyWaitGroup *sync.WaitGroup
+}
+
+func (c *chaosMonkeyMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if err := util.ToStruct(config.Params, &c.NodeFailureConfig); err != nil {
+			return nil, err
+		}
+		c.stopChannel = make(chan struct{})
+		c.chaosMonkey = chaos.NewMonkey(config.ClusterFramework.GetClientSets().GetClient(), config.CloudProvider)
+		c.chaosMonkeyWaitGroup, err = c.chaosMonkey.Init(api.ChaosMonkeyConfig{&c.NodeFailureConfig}, c.stopChannel)
+		if err != nil {
+			close(c.stopChannel)
+			return nil, fmt.Errorf("error while creating chaos monkey: %v", err)
+		}
+		return nil, nil
+	case "stop":
```

Comment:
Is it possible to add case "pause", "resume" (even as a follow up)?

NodeKiller ensures that single VM will be restarted once. Running multiple measurements disables this feature.

Reply:
Pretty sure it's possible. Not sure how complicated it would be, given that current chaos monkey implementation doesn't expect such usage.

Is there any reason to enforce single VM restart throughout the test? It seems like it weakens the test to introduce this assumption, but I don't have much context here.

Reply:
It's related to restart count metric. With enabled chaos monkey we are ok with single restart of each component on each node. Current configuration overwrites this (see https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/chaosmonkey/ignore_node_killer_container_restarts_100.yaml).

We need to match ignored restart to numbers in this configuration, otherwise we will alert on restarting components.

Alternatively we could incorporate connection between Chaos Monkey and restart count measurement to discover and ignore restart generated by Chaos Monkey, but it seems like a much more work to do.